crates/miden-lib/src/account/interface/mod.rs

```diff
@@ -0,0 +1,360 @@
+    }
+
+    // probably it would be convenient to also have `from_components` constructor
+}
+
+// ACCOUNT COMPONENT INTERFACE
+// ================================================================================================
+
+/// Describes whether a note is compatible with a specific account.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum NoteAccountCompatibility {
+    /// A note is incompatible with an account.
+    ///
+    /// The account interface does not have procedures for being able to execute at least one of
+    /// the program execution branches.
+    No,
+    /// The account has all necessary procedures of one execution branch of the note script. This
+    /// means the note may be able to be consumed by the account if that branch is executed.
+    Maybe,
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Returns a vector of digests of all account component interfaces.
+pub fn component_proc_digests(
+    account_component_interfaces: &[AccountComponentInterface],
+) -> Vec<Digest> {
+    let mut component_proc_digests = Vec::new();
+    for component in account_component_interfaces.iter() {
+        match component {
+            AccountComponentInterface::BasicWallet => {
+                component_proc_digests
+                    .extend(basic_wallet_library().mast_forest().procedure_digests());
+            },
+            AccountComponentInterface::BasicFungibleFaucet => {
+                component_proc_digests
+                    .extend(basic_fungible_faucet_library().mast_forest().procedure_digests());
+            },
+            AccountComponentInterface::RpoFalcon512(_) => {
+                component_proc_digests
+                    .extend(rpo_falcon_512_library().mast_forest().procedure_digests());
+            },
+            AccountComponentInterface::Custom(custom_procs) => {
+                component_proc_digests.extend(custom_procs.iter().map(|info| *info.mast_root()));
+            },
+        }
+    }
+
+    component_proc_digests
+}
+
+/// Verifies that the provided note script is compatible with the target account interfaces.
+///
+/// This is achieved by checking that at least one execution branch in the note script is compatible
+/// with the account procedures vector.
+///
+/// This check relies on the fact that account procedures are the only procedures that are `call`ed
+/// from note scripts, while kernel procedures are `sycall`ed.
+fn verify_note_script_compatibility(
+    note_script: &NoteScript,
+    account_procedures: Vec<Digest>,
+) -> NoteAccountCompatibility {
+    // collect call branches of the note script
+    let branches = collect_call_branches(note_script);
+
+    // if none of the branches are compatible with the target account, return a `CheckResult::No`
+    if !branches
+        .iter()
+        .any(|call_targets| call_targets.iter().all(|target| account_procedures.contains(target)))
```

Comment:
If we work with `BTreeSet`s, we may be able to use something like `BTreeSet::is_subset()` here.